[Refactor] OpeningSplashView moved from screensaver to view file

FazleRabbbiferdaus172 opened the pull request:
## Description

This PR refactors the code by moving the `OpeningSplash`,  `LogoScreen`, `OpeningSplashScreen`, `ScreensaverScreen`  class to its corresponding view file, improving the organization of the codebase. Additionally, the imports are updated to reflect this change and refactored `ScreensaverScreen`.

**Changes made:**
- Moved `OpeningSplash` class to the appropriate view file.
- Fixed the imports after moving the class.
- In the flow test base , `run `method of `OpeningSplashView` is mocked.
- Moved `LogoScreen` class to `gui.screens.screen`.
- Moved `OpeningSplashScreen`  class to `gui.screens.screen`.
- Moved `ScreensaverScreen` class to `gui.screens.screen`.
- Remove `buttons` parameter from `ScreensaverScreen` class.
- Deleted `views.screensaver`.

![Screenshot from 2025-04-09 00-23-09](https://github.com/user-attachments/assets/b6d33e3f-36bf-4e7a-af54-257310b5e440)

This pull request is categorized as a:

- [ ] New feature
- [ ] Bug fix
- [x] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [x] N/A

I have tested this PR on the following platforms/os:

- [ ] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [x] Other

FazleRabbbiferdaus172 committed:
[Refactor] OpeningSplashView moved from screensaver to view file

OpeningSplash view seems more appropriate in the view file. imports are fixed after moving the class. Base test should mock the run method of OpeningSplashView as it was mocked before but instead of mocking the whole module only the run method is mocked to properly run the test.

kdmukai commented on tests/base.py:
```diff
@@ -181,121 +180,122 @@ def run_sequence(self, sequence: list[FlowStep], initial_destination_view_args:
-                        if cur_flow_step.screen_return_value is None and cur_flow_step.button_data_selection is None:
-                            # This is the last View in the sequence and it doesn't specify any
-                            # user-mimicking interactions for the Screen. Nothing left to do.
+        with patch.object(OpeningSplashView, 'run', return_value=None):
```
This would be better handled as a `MagicMock` as it had previously been for the whole `screensaver` module.

I'd have to experiment with exactly what to `MagicMock` around this new `OpeningSplashView` to get it to work. But there should be some way to do it.

This extra `patch` indent just feels way too ugly. And github's visual diff here is bad at displaying the changes, making it hard to review this without pulling the PR locally (for the moment I'm only doing high-level, quick-ish glance PR reviews. Don't yet have time to go more in depth and run PRs locally).

FazleRabbbiferdaus172 replied:
I've updated the code to use MagicMock for OpeningSplashView.run as suggested. This simplifies the test setup and avoids the extra patch indentation. Let me know if there's anything else you'd like adjusted!

kdmukai commented:
I like the basics of this fairly simple change.

But I'd say it's only part of the way there:
* Refactor the `View` into a better location: ✅
* Refactor its associated `Screen` into a better location: ?
* Fix up the test suite harness in a clean(er) manner: ?
* Ensure that the screenshot generator can still generate the opening splash screen: (might depend on mocking strategy)?

FazleRabbbiferdaus172 committed:
[Refactor] Use MagicMock for OpeningSplashView.run instead of patching

Replace patch.object with MagicMock for OpeningSplashView.run to simplify test setup

FazleRabbbiferdaus172 commented:
Thanks for the feedback!

- I’ve updated the code to use MagicMock instead of patch.object for OpeningSplashView.run, which simplifies the test harness.

- I’ve verified that the screenshot generator is working as expected—all the locale variants and the Opening Splash screen are being generated correctly.

- I have been studying the code base and I’d also love to proceed with refactoring the associated Screen into gui.screens.screen, and introducing a ScreensaverView to better align with the project’s structure. That refactor would include breaking out and reorganizing the logic currently in LogoScreen and ScreensaverScreen.

Let me know if you'd prefer those screen/view refactors to happen in this PR or a follow-up one!

FazleRabbbiferdaus172 commented:
screenshot references

![image](https://github.com/user-attachments/assets/fb0608c0-749e-40be-835b-c28fa724ae21)

![image](https://github.com/user-attachments/assets/889f375d-30d9-452e-b86d-139fb3c96d06)

kdmukai commented:
> * I’d also love to proceed with refactoring the associated Screen into gui.screens.screen, and introducing a ScreensaverView to better align with the project’s structure. That refactor would include breaking out and reorganizing the logic currently in LogoScreen and ScreensaverScreen.

Note my explanation here: https://github.com/SeedSigner/seedsigner/pull/729#issuecomment-2799992010

But overall it sounds like you're on the right track and, yes, I'd like to see the above restructure / refactor.

FazleRabbbiferdaus172 commented:
Appreciate the feedback! I’ll continue working on the restructure and refactor in this PR. In the meantime, I’m also looking into the TODO for the disappearing ToastOverlay mentioned in the ScanScreen to get a better grasp of the code internals. Once I wrap that up, I’ll shift focus back to this.

FazleRabbbiferdaus172 committed:
[Refactor] Move screens to gui/screens/screen.py

Moved LogoScreen, OpeningSplashScreen, and ScreensaverScreen from views/screensaver.py
to gui/screens/screen.py to align with the current project structure and improve
modularity. This change helps maintain consistency by keeping all screen-related
components under gui/screens/.

No functional changes were introduced.

FazleRabbbiferdaus172 commented:
I've pushed a new commit that refactors and moves LogoScreen, OpeningSplashScreen, and ScreensaverScreen from views/screensaver.py to gui/screens/screen.py to match the current project structure. No functional changes were made.

kdmukai commented:
Minor note @FazleRabbbiferdaus172: update the PR description to include the `Screen` changes and anything else that has been added/altered.

kdmukai commented on src/seedsigner/controller.py:
```diff
@@ -13,7 +13,7 @@
 from seedsigner.models.singleton import Singleton
 from seedsigner.models.threads import BaseThread
-from seedsigner.views.screensaver import ScreensaverScreen
+from seedsigner.gui.screens import ScreensaverScreen
```
Well... looking at the diff again I see that import was already up there to begin with before these changes...

Hrmm... The second import on line 392 would be redundant then. I'd still lean toward removing the top-level import and just keep the later one.

kdmukai commented on tests/base.py:
```diff
@@ -1,26 +1,27 @@
-import logging
 logger = logging.getLogger(__name__)
 
+OpeningSplashView.run = MagicMock()
```
Can you add a comment here to explain why this is necessary?

FazleRabbbiferdaus172 replied:
added the comment

kdmukai commented on tests/screenshot_generator/generator.py:
```diff
@@ -38,7 +38,7 @@
 from seedsigner.views import (MainMenuView, PowerOptionsView, RestartView, NotYetImplementedView, UnhandledExceptionView, 
     psbt_views, seed_views, settings_views, tools_views, scan_views)
-from seedsigner.views.screensaver import OpeningSplashView
+from seedsigner.views.view import OpeningSplashView
```
This import can now be combined into a single line with the import below.

FazleRabbbiferdaus172 replied:
combined the imports

kdmukai commented on src/seedsigner/controller.py:
```diff
@@ -389,7 +389,7 @@ def start_screensaver(self):
-            from seedsigner.views.screensaver import ScreensaverScreen
+            from seedsigner.gui.screens import ScreensaverScreen
             from seedsigner.hardware.buttons import HardwareButtons
             self.screensaver = ScreensaverScreen(HardwareButtons.get_instance())
```
While you're here, suggested minor refactor: there's no reason to pass the `HardwareButtons` into the `ScreensaverScreen` anymore; this is an outdated artifact from much older code before `HardwareButtons` was refactored into a singleton.

`ScreensaverScreen` can retrieve the `HardwareButtons` instance on its own.

FazleRabbbiferdaus172 replied:
Remove buttons parameter from ScreensaverScreen and made the necessary

FazleRabbbiferdaus172 committed:
[REF] Minor refactor of comments and import cleanup

FazleRabbbiferdaus172 committed:
[Refactor] Remove HardwareButtons parameter from ScreensaverScreen

Since HardwareButtons is already accessible via self.hw_inputs in BaseScreen, which ScreensaverScreen inherits through LogoScreen, there is no need to pass it explicitly from the controller.

FazleRabbbiferdaus172 commented:
Made the change according to review and updated the pr description

FazleRabbbiferdaus172 commented:
I will resolve the conflicts and update the current branch to dev, as soon as I can

FazleRabbbiferdaus172 committed:
[Refactor] moved changes from views/screensaver to gui/screens/screen. views/screensaver is removed.

FazleRabbbiferdaus172 committed:
[REF] buttons argument removed from ScreenSaverScreen

as hw_inputs is already a property of base screen. Passing hardware button instance in no longer needed.

FazleRabbbiferdaus172 committed:
[REF] Add dedicated ScreenSaverView component

to meet the current pattern of view and screen ScreenSaverView is introduced.

newtonick commented:
I'm not sure of the cause, but when testing this PR, the splash does not dismiss correctly. I've compared the behavior to the current dev branch (7dd2e07) and there is a difference. The "You can remove the SD card now" message remains on the screen and is not wiped before the home screen gui components are redrawn after the screensaver stops and restores the last screen.

PR 718 behavior as of 4899f61 (with gui artifact issue)

https://github.com/user-attachments/assets/8537e389-583b-421f-bf11-8f21beadbbb1

dev branch behavior (without gui artifact issue)

https://github.com/user-attachments/assets/55f4a98b-2adc-45aa-8d5b-73ac4c70e422

FazleRabbbiferdaus172 committed:
screensaver run method should acquire and release lock, Otherwise the toast threads will not require the lock after the screensaver stops.

FazleRabbbiferdaus172 commented:
> 'm not sure of the cause, but when testing this PR, the splash does not dismiss correctly. I've compared the behavior to the current dev branch ([7dd2e07](https://github.com/SeedSigner/seedsigner/commit/7dd2e075e0e6ef5228ed7e161c4e2c43adfac133)) and there is a difference. The "You can remove the SD card now" message remains on the screen and is not wiped before the home screen gui components are redrawn after the screensaver stops and restores the last screen.
> 
> PR 718 behavior as of [4899f61](https://github.com/SeedSigner/seedsigner/commit/4899f61844e55014587a5b8082f6bbd520c249ad) (with gui artifact issue)

found the issue, In the refactor screensaver screen `_run` was not grabbing the render lock and BaseToastOverlayManagerThread run method expects some other process to lock the renders and only then the BaseToastOverlayManagerThread will attempt to reacquire the lock for it's self. also BaseScreen _run method doc string also mentions that any class that implements _run should acquire the renderer.lock. I fixed the issue and test in an emulator #[05d60e9](https://github.com/SeedSigner/seedsigner/pull/718/commits/05d60e9681360f00838ee873a96d27266478a7dd)

kdmukai commented on src/seedsigner/views/view.py:
```diff
@@ -183,6 +183,55 @@ def __ne__(self, obj):
+@dataclass
+class ScreenSaverView(View):
+    def __post_init__(self):
+        self._is_running = False
```
It feels like unnecessary duplication to track another `_is_running` state var when the `ScreensaverScreen` instance only exists when it's running.

```suggestion
        self.screensaver_screen = None
        
        ...
        
    @property
    def is_running(self):
        return self.screensaver_screen is not None
```

Obv we'd set the var in `run()`.

And then set it back to `None` on `stop()`. And maybe also at the end of `run()`? I'm not sure it makes a difference there.

FazleRabbbiferdaus172 replied:
The `View` class already has a `screen` attribute that is set when run_screen is called. Can we reuse this existing attribute instead of introducing a new `screensaver_screen` attribute as you suggested?

kdmukai commented on tests/screenshot_generator/generator.py:
```diff
@@ -16,7 +16,7 @@
 sys.modules['seedsigner.hardware.displays.st7789_mpy'] = MagicMock()
 sys.modules['seedsigner.hardware.displays.ili9341'] = MagicMock()
-sys.modules['seedsigner.views.screensaver.ScreensaverScreen'] = MagicMock()
+sys.modules['seedsigner.gui.screens.ScreensaverScreen'] = MagicMock()
```
Because `ScreensaverScreen` is now being called by `run_screen` in its View, it doesn't need to be mocked out. The screenshot generator already prevents any Screen from being instantiated in `run_screen`.

kdmukai commented on src/seedsigner/controller.py:
```diff
@@ -388,13 +388,10 @@ def start_screensaver(self):
         
         # Start the screensaver, but it will block until it can acquire the Renderer.lock.
-        self.screensaver.start()
+        self.screensaver.run()
```
This is a minor nit, but I would prefer a `ScreensaverView.start()` method to parallel its `stop()` method.

```python
def start(self):
    self.run()
```

kdmukai commented on src/seedsigner/gui/screens/screen.py:
```diff
@@ -1315,6 +1317,217 @@ def update_title(self) -> bool:
+        self.increment_y = self.rand_increment()
+
+        self._is_running = False
+        # Store the current screen in order to restore it later
+        self.last_screen = self.renderer.canvas.copy()
```
VERY minor subtlety here. This `last_screen` was originally set on Screensaver `start()`.

Now that this is all in a proper Screen class, we have the standard `__init__()` vs `_run()` separation, HOWEVER that all happens on Screensaver `start()` anyway. So it's totally fine to set `last_screen` in the `__init__()` here.

It just FEELS a tiny bit worrisome to set it in `__init__()` instead of in `_run()`. IF you don't read the rest of the code to see how this Screen is used, you might wonder: "Wait, if the `last_screen` is set when it's instantiated, couldn't the state be different when the screensaver is actually `_run()`?"

Also note that `last_screen` is only ever referenced within `_run()`.

SO... for those reasons, I would prefer that it NOT be a member var set in `__init__()` and instead just be a var that is set within and is only local to `_run()`.

FazleRabbbiferdaus172 replied:
The `clear_screen` method gets called before `_run` method, which wipes the canvas clean. Consequently, if we try to capture the `last_screen` within the `_run` method, we will only get a blank image. That is why I moved `last_screen` in init as the `run_screen` creates a new screen object.

To solve this, we can make `last_screen` a local variable and override the `_render` method to prevent `clear_screen` from being called, ensuring the canvas is not cleared before we capture it.

I can make the required changes if you want.

kdmukai replied:
> The `clear_screen` method gets called before `_run` method, which wipes the canvas clean. Consequently, if we try to capture the `last_screen` within the `_run` method, we will only get a blank image. That is why I moved `last_screen` in init as the `run_screen` creates a new screen object.

Ah, wonderfully insightful! I would add a brief explanation of this as a comment then for why it's set in the `__init__()`.

Though I'm just quickly reviewing now, I'll need to dig deeper later to make sure I fully grasp what's happening.

FazleRabbbiferdaus172 replied:
added a comment that explain why is necessary.

FazleRabbbiferdaus172 committed:
[REF] start method is favoured over run

FazleRabbbiferdaus172 committed:
[REF] ScreensaverScreen does not need to be mocked

FazleRabbbiferdaus172 committed:
comment added that explains why last_screen should be captured in __init__
